Enhance the UX for packages with License Expressions

nkolev92 opened the pull request:
## Bug
Fixes: https://github.com/NuGet/Home/issues/7391
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 
Add license expression details to the Details and License Acceptance page. 

![nonstandardlicense](https://user-images.githubusercontent.com/2878341/46924033-3214ae80-cfd5-11e8-9fc1-f1dba7cd31c9.png)
![complexlicense](https://user-images.githubusercontent.com/2878341/46924034-3214ae80-cfd5-11e8-9a97-c01b00e950df.png)
![complexnonstandardlicense](https://user-images.githubusercontent.com/2878341/46924035-32ad4500-cfd5-11e8-86dd-4f3ce4d5c8e5.png)
![requirelicenseacceptance](https://user-images.githubusercontent.com/2878341/46924036-32ad4500-cfd5-11e8-9f48-b669a11e5fc7.png)

## Testing/Validation
Tests Added: Yes, where appropriate.  
Reason for not adding tests:  
Validation done:

nkolev92 committed:
Initial license expression UIwork

nkolev92 committed:
add warning icon from the known monikers

nkolev92 committed:
fix margins

nkolev92 committed:
work

nkolev92 committed:
work

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs:
```diff
@@ -19,6 +19,9 @@
 using Task = System.Threading.Tasks.Task;
+using NuGet.Packaging;
+using NuGet.Packaging.Licenses;
+using System.Collections.Specialized;
```
fix ordering

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs:
```diff
@@ -0,0 +1,133 @@
+            }
+            else
+            {
+                if (licenseUrl != null)
```
combined into a `else if`

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs:
```diff
@@ -0,0 +1,133 @@
+            return list;
+        }
+
+        private static void GetLicenseIdentifiers(NuGetLicenseExpression expression, IList<string> identifiers)
```
either return the identifier list or change the name to populateLicenseIdentifier

nkolev92 replied:
Changed it to populate.

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs:
```diff
@@ -0,0 +1,133 @@
+                                list.Add(new FreeText(licenseToBeProcessed.Substring(0, licenseStart)));
+                            }
+                            var license = licenseToBeProcessed.Substring(licenseStart, identifier.Length);
+                            list.Add(new LicenseText(license, new Uri($"https://spdx.org/licenses/{license}.html")));
```
create a constant for the spdx url

nkolev92 replied:
I was gonna change this before merging. 
It'll be our Url, I will create a constant for it regardless though :)

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs:
```diff
@@ -0,0 +1,133 @@
+                    break;
+                case LicenseType.File:
+
+                    list.Add(new FreeText($"License '{metadata.License}' is embedded in the package."));
```
add a resource string and do we need to localize this?

nkolev92 replied:
Added

jainaashish commented on src/NuGet.Clients/NuGet.Tools/NuGetUIFactory.cs:
```diff
@@ -61,7 +61,8 @@ internal sealed class NuGetUIFactory : INuGetUIFactory
             ICommonOperations commonOperations,
             INuGetUILogger logger,
-            ISourceControlManagerProvider sourceControlManagerProvider)
+            ISourceControlManagerProvider sourceControlManagerProvider,
```
skip this from your PR. Pato will take care of that

nkolev92 replied:
I have merged Pato's PR, will rebase from dev.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/NuspecReader.cs:
```diff
@@ -471,7 +471,29 @@ public LicenseMetadata GetLicenseMetadata()
+                                            invalidLicenseIdentifiers.Add(nugetLicense.Identifier);
+                                        }
+                                    };
+                                    expression.OnEachLeafNode(licenseProcessor, null);
```
its so hard to read what exactly you're trying to do here...

nkolev92 replied:
I will add some comments. 
But basically this is a visitor for each leaf node, and if it sees any non standard licenses it creates a warning message.

jainaashish commented on test/TestUtilities/Test.Utility/JsonData.cs:
```diff
@@ -4651,5 +4651,111 @@ public static class JsonData
+            ""iconUrl"": """",
+            ""id"": ""PackageA"",
+            ""language"": """",
+            ""licenseUrl"": ""https://github.com/PackageA-contrib/PackageA-hosting/blob/master/LICENSE.txt"",
```
Would server also return license expression in search metadata for newer clients? or we always going to read it from nuspec file explicitly to show in browse tab?

nkolev92 replied:
Yes, the server will return a licenseExpression as per the design of @agr.

jainaashish reviewed (approved):
Few small comments, overall looks good.

nkolev92 committed:
work

nkolev92 committed:
fix compilation error

nkolev92 committed:
refacotr

nkolev92 committed:
fix compilation error

nkolev92 committed:
fix test

nkolev92 committed:
fix npe

PatoBeltran commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs:
```diff
@@ -0,0 +1,45 @@
+
+    internal class LicenseText : IText
+    {
+        public LicenseText(string text, Uri link)
```
can `text` and `link` be null? If not, add null checks

PatoBeltran commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs:
```diff
@@ -0,0 +1,45 @@
+        string Text { get; }
+    }
+
+    internal class LicenseText : IText
```
I would suggest to have `LicenseText`, `FreeText` and `WarningText` in their own file for better discovery of the code. I don't have an strong opinion, so if you feel strongly against it disregard this comment.

nkolev92 replied:
I'll move them to separate files.

PatoBeltran commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        public FreeText(string text)
+        {
+            Text = text;
```
add null check if needed

PatoBeltran commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        public WarningText(string text)
+        {
+            Text = text;
```
add null check if needed

PatoBeltran commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs:
```diff
@@ -0,0 +1,129 @@
+                    if (metadata.LicenseExpression != null)
+                    {
+                        IList<string> identifiers = new List<string>();
+                        PopulateLicenseIdentifiers(metadata.LicenseExpression, ref identifiers);
```
if you are only adding you don't need to have the param as `ref`

nkolev92 replied:
good catch.
I think initially I was assigning that reference inside.

PatoBeltran commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs:
```diff
@@ -0,0 +1,129 @@
+
+                    if (metadata.LicenseExpression != null)
+                    {
+                        IList<string> identifiers = new List<string>();
```
nit: `var identifiers`

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/NuspecReader.cs:
```diff
@@ -432,7 +432,7 @@ public LicenseMetadata GetLicenseMetadata()
                 var isKnownType = Enum.TryParse(type, ignoreCase: true, result: out LicenseType licenseType);
 
-                IList<string> errors = null;
+                List<string> errors = null;
```
why change this from `IList` to `List`?

nkolev92 replied:
I changed the warnings and errors to IReadOnlyList from IList. 

And List implements both :)

PatoBeltran reviewed (commented):
I think the warning icon should be left of the text. This way if its consistent with most icons in VS and also if the text is really big, and the icon is right of the text it might get hidden and make the user need to scroll to see it.

nkolev92 committed:
address feedback

nkolev92 commented:
@PatoBeltran 
Addressed your feedback. 

Moved the warning in front of the text.
![warningsignfirst](https://user-images.githubusercontent.com/2878341/47040347-e3812480-d13a-11e8-9e63-ba631ce32d28.png)

Can you peek at it again.

nkolev92 commented:
All the tests failing are due to the repo signing.

nkolev92 committed:
add better test coverage

nkolev92 committed:
fix tests
